### ISSUE TITLE
False positive for PERF401 when conditional depends on list being appended

### ISSUE THREAD

**Issue by dosisod** [astral-sh/ruff#5581]
The following code cannot be re-written as a comprehension. It's a pretty common idiom (from my experience) to remove duplicates from an array while maintaining the order, keeping the first duplicate.

```python
copy: list[int] = []

for i in [1, 1, 2, 2, 3, 3]:
    if i not in copy:
        copy.append(i)

print(copy)  # [1, 2, 3]
```

Since `copy` is being appended to, you cannot use it the conditional branch of a comprehension:

```python
copy = [i for i in [1, 1, 2, 2, 3, 3] if i not in copy]  # error, copy is not defined
```

If `i` is a hashable type, you could use this instead of a comprehension:

```python
nums = [1, 1, 2, 2, 3, 3]
copy = list(dict.fromkeys(nums))

print(copy)  # [1, 2, 3]
```

Though if you're iterating over non-hashable types (which is the case for me), `dict.fromkeys()` won't work.

Ruff output from first example:

```
$ ruff file.py
file.py:5:9: PERF401 Use a list comprehension to create a transformed list
```

Ruff version:

```
$ ruff --version
ruff 0.0.277
```

**Comment by charliermarsh** [comment]
Hmm, makes sense. We can fix this.
